packages/plugin-ext/src/main/browser/view/tree-views-main.tsx

```diff
@@ -281,11 +375,50 @@ export class TreeViewWidget extends TreeWidget {
     }
 
     renderIcon(node: TreeNode, props: NodeProps): React.ReactNode {
+
         if (node.icon) {
             return <div className={'fa ' + node.icon + ' tree-view-icon'}></div>;
+        } else {
+            const nodeDMD = node as DescriptiveMetadata;
+
+            if (nodeDMD && nodeDMD.metadata && this.metadata) {
+                const iconPath = nodeDMD.metadata.iconPath;
+                let lightIconUrl = typeof iconPath === 'object' ? iconPath.light : iconPath;
+                const iconClass = this.createClassName(lightIconUrl, this.metadata.packagePath);
+
+                // insert dynamic css class
+                if (!this.style.exists('.' + iconClass)) {
+                    let darkIconUrl = typeof iconPath === 'object' ? iconPath.dark : iconPath;
+                    darkIconUrl = this.getIconUrl(darkIconUrl, this.metadata.packagePath, this.metadata.urlBase);
+                    lightIconUrl = this.getIconUrl(lightIconUrl, this.metadata.packagePath, this.metadata.urlBase);
+
+                    this.style.insertRule('.' + iconClass, theme => `
+                        width: 22px;
+                        height: 22px;
+                        background: no-repeat url("${theme.id === BuiltinThemeProvider.lightTheme.id ? lightIconUrl : darkIconUrl}");
+                    `);
+                }
+                return <div className={iconClass + ' tree-view-icon'}></div>;
+            }
         }
+        return '';
+    }
 
-        return undefined;
+    private createClassName(iconPath: string, packagePath: string): string {
+        if (iconPath.startsWith(packagePath)) {
+            iconPath = iconPath.substr(packagePath.length);
+        }
+        return iconPath.replace(/\//g, '_').replace(/\./g, '_');
+    }
+
+    private getIconUrl(iconPath: string, packagePath: string, urlBase: string): string {
+        if (iconPath.startsWith(packagePath)) {
```

Comment:
there is `URI.relative`

Reply:
It seems to fit http uri, with scheme check and authority check. Here it is a filesystem path.